### DISCUSSION TITLE
Permission Denied from query

### DISCUSSION THREAD

**Discussion by bunkiedc** [ClickHouse/ClickHouse#58840] [Q&A]
I'm getting the following error when submitting a query via clickhouse-client via a script:

Code: 83. DB::Exception: Couldn't create directory /user_defined/ reason: 'Permission denied': While executing ValuesBlockInputFormat: data for INSERT was parsed from query: (in query: insert into customer_info values(blah blah blah...). (DIRECTORY_DOESNT_EXIST)

My environment is running clickhouse server in a container.  This error just started, as we had a server blow out and we're attempting to recover everything.  I can do a query insert from either clickhouse or via the command line fine.  I know the directory exists as I can see it under /var/lib/docker/volumes/jeda_clickhouse/_data/user_defined, it just looks like something can't find this directory?  The container is also running as root.  We aren't using any user defined functions, so I'm not quite sure why it's being checked/written to.
thanks,
          david

**Comment by bunkiedc** [comment]
For some reason, running the script at root for 1 time, cleared the permission issue.  I have no idea why, but is works as expected now.

**Comment by bunkiedc** [comment]
We are a startup just about ready to release our product, but this is keeping us from doing so.  Any help would be greatly appreciated!!

This problem has come back.  I've tried updating both clickhouse-client and clickhouse-server to the latest to no avail.  The permission denied only appears when script 1 calls script 2, that does an insert via clickhouse-client (i.e. script 2 contains the clickhouse-client command).  However, running script 2 from the command line (as the same user) does not get the permission denied. Putting the same clickhouse-client command directly into script 1 (i.e. not calling script 2) **does** get the permission denied error. As I said, both scripts are running via the same user.

I have also turned trace level on for logging, but didn't get any additional useful information.

In **all** cases that **do not** get the permission denied error, the "user_defined" directory is getting created in the home directory of the server where the insert is being run from.  If the "user_defined" directory is removed from the real path (/var/lib/clickhouse/user_defined), clickhouse-server **does not** recreate it there, only in my home directory.  I assume that the two events (permission denied and creating the directory in the wrong location) are related, but I am unable to prove this as I can't tell further where clickhouse thinks it is (i.e. I'm not seeing it in the trace logging).

ClickHouse client version 24.1.3.31 (official build).
Connected to ClickHouse server version 24.1.2.

I've checked the paths of clickhouse-server from the preprocessed_configs/config.xml file.

<ssl_cert_file>/path/to/ssl_cert_file</ssl_cert_file>
<ssl_key_file>/path/to/ssl_key_file</ssl_key_file>
<ssl_ca_cert_file>/path/to/ssl_ca_cert_file</ssl_ca_cert_file>

**<path>/var/lib/clickhouse/</path>**
<path>/data/</path>
<endpoint>http://path/to/endpoint </endpoint>
<metadata_path>/var/lib/clickhouse/disks/blob_storage_disk/</metadata_path>

<tmp_path>/var/lib/clickhouse/tmp/</tmp_path>
If not set <tmp_path> is used, otherwise <tmp_path> is ignored.
<user_files_path>/var/lib/clickhouse/user_files/</user_files_path>
tls_cert_file - path to certificate file.
tls_key_file - path to certificate key file.
tls_ca_cert_file - path to CA certificate file.
tls_ca_cert_dir - path to the directory containing CA certificates.
<tls_cert_file>/path/to/tls_cert_file</tls_cert_file>
<tls_key_file>/path/to/tls_key_file</tls_key_file>
<tls_ca_cert_file>/path/to/tls_ca_cert_file</tls_ca_cert_file>
<tls_ca_cert_dir>/path/to/tls_ca_cert_dir</tls_ca_cert_dir>
tls_cert_file - path to certificate file.
tls_key_file - path to certificate key file.
tls_ca_cert_file - path to CA certificate file.
tls_ca_cert_dir - path to the directory containing CA certificates.
<tls_cert_file>/path/to/tls_cert_file</tls_cert_file>
<tls_key_file>/path/to/tls_key_file</tls_key_file>
<tls_ca_cert_file>/path/to/tls_ca_cert_file</tls_ca_cert_file>
<tls_ca_cert_dir>/path/to/tls_ca_cert_dir</tls_ca_cert_dir>

<path>users.xml</path>

<path>/var/lib/clickhouse/access/</path>

By default, path to file with substitutions is /etc/metrika.xml. It could be changed in config in 'include_from' element.
root_path - prefix for keys
hostname_in_path - append hostname to root_path (default = true)
<root_path>one_min</root_path>
<hostname_in_path>true</hostname_in_path>
<root_path>one_sec</root_path>
endpoint - mertics path (relative to root, statring with "/")

Format: /path/to/file
Path to the list is under top_level_domains_path (see above).
<public_suffix_list>/path/to/public_suffix_list.dat</public_suffix_list>
<public_suffix_list>/path/to/public_suffix_list.dat</public_suffix_list>

<path>/clickhouse/jeda_cluster/task_queue/ddl</path>

If you want do delete one table and don't want to change clickhouse-server config
<details><summary>Details</summary>
</details> , you could create special file /flags/force_drop_table and make DROP once.

If not set <tmp_path> is used, otherwise <tmp_path> is ignored.
By default, path to file with substitutions is /etc/metrika.xml. It could be changed in config in 'include_from' element.
root_path - prefix for keys
hostname_in_path - append hostname to root_path (default = true)
<root_path>one_min</root_path>
<hostname_in_path>true</hostname_in_path>
<root_path>one_sec</root_path>
endpoint - mertics path (relative to root, statring with "/")

Format: /path/to/file
Path to the list is under top_level_domains_path (see above).
If you want do delete one table and don't want to change clickhouse-server config, you could create special file /flags/force_drop_table and make DROP once.
<!-- Directory in containing schema files for various input formats.

**Comment by bunkiedc** [comment]
Can someone help with this or provide a suggestion to see why/where clickhouse is doing this?  It looks like the directory creation can also happen in whatever directory the script is called from.  I've now found these directories in multiple places.

**Comment by ArctypeZach** [comment]
Hey @bunkiedc - can you please open an issue for this?

**Comment by bunkiedc** [comment]
Hi @ArctypeZach, there is an issue opened for this... [59597](https://github.com/ClickHouse/ClickHouse/issues/59597). Do I need to do something more??
